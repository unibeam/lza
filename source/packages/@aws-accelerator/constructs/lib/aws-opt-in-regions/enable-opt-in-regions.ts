/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import path = require('path');

/**
 * Opt-in Regions Props
 */
export interface OptInRegionsProps {
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
  /**
   * Custom resource lambda management account id
   */
  readonly managementAccountId: string;
  /**
   * Custom resource lambda log account ids
   */
  readonly accountIds: string[];
  /**
   * Custom resource lambda home region
   */
  readonly homeRegion: string;
  /**
   * Custom resource lambda enabled regions
   */
  readonly enabledRegions: string[];
  /**
   * Custom resource lambda global region
   */
  readonly globalRegion: string;
}

/**
 * Class Opt-in Regions
 */
export class OptInRegions extends Construct {
  readonly onEvent: cdk.aws_lambda.IFunction;
  readonly isComplete: cdk.aws_lambda.IFunction;
  readonly provider: cdk.custom_resources.Provider;
  readonly id: string;

  constructor(scope: Construct, id: string, props: OptInRegionsProps) {
    super(scope, id);

    const OPT_IN_REGIONS = 'Custom::OptInRegions';

    const AccountOperationsPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'AccountOperations',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['account:ListRegions', 'account:EnableRegion', 'account:GetRegionOptStatus'],
      resources: ['*'],
    });

    const serviceAccessAndTokenPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'serviceAccessAndToken',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: [
        'organizations:EnableAWSServiceAccess',
        'iam:GetAccountSummary',
        'iam:SetSecurityTokenServicePreferences',
      ],
      resources: ['*'],
    });

    this.onEvent = new cdk.aws_lambda.Function(this, 'OptInRegionsOnEvent', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'enable-opt-in-regions/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(1),
      description: 'Opt-in Regions onEvent handler',
      environmentEncryption: props.kmsKey,
      initialPolicy: [serviceAccessAndTokenPolicy],
    });

    const onEventLogGroup = new cdk.aws_logs.LogGroup(this, `${this.onEvent.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${this.onEvent.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.kmsKey,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    this.isComplete = new cdk.aws_lambda.Function(this, 'OptInRegionsIsComplete', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'enable-opt-in-regions-status/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(5),
      description: 'Opt-in Regions isComplete handler',
      environmentEncryption: props.kmsKey,
      initialPolicy: [AccountOperationsPolicy],
    });

    const isCompleteLogGroup = new cdk.aws_logs.LogGroup(this, `${this.isComplete.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${this.isComplete.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.kmsKey,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    this.provider = new cdk.custom_resources.Provider(this, 'OptInRegionsProvider', {
      onEventHandler: this.onEvent,
      isCompleteHandler: this.isComplete,
      queryInterval: cdk.Duration.seconds(120),
      totalTimeout: cdk.Duration.minutes(30),
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: OPT_IN_REGIONS,
      serviceToken: this.provider.serviceToken,
      properties: {
        props: {
          managementAccountId: props.managementAccountId,
          accountIds: props.accountIds,
          homeRegion: props.homeRegion,
          enabledRegions: props.enabledRegions,
          globalRegion: props.globalRegion,
        },
      },
    });

    // Ensure that the LogGroup is created by Cloudformation prior to Lambda execution
    resource.node.addDependency(isCompleteLogGroup);
    resource.node.addDependency(onEventLogGroup);
    this.id = resource.ref;
  }
}
