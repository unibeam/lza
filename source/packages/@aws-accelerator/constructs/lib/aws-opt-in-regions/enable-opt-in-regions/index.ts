/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/**
 * aws-opt-in-regions - lambda handler
 *
 * @param event
 * @returns
 */
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import { setStsTokenPreferences } from '@aws-accelerator/utils/lib/set-token-preferences';
import { OrganizationsClient, EnableAWSServiceAccessCommand } from '@aws-sdk/client-organizations';
import { setRetryStrategy } from '@aws-accelerator/utils/lib/common-functions';

const solutionId: string = process.env['SOLUTION_ID'] ?? '';

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      IsComplete: boolean;
    }
  | undefined
> {
  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const { accountIds, globalRegion } = event.ResourceProperties['props'] as {
        accountIds: string[];
        globalRegion: string;
      };

      // enable trust access
      const organizationsClient = new OrganizationsClient({
        customUserAgent: solutionId,
        retryStrategy: setRetryStrategy(),
      });
      try {
        const command = new EnableAWSServiceAccessCommand({
          ServicePrincipal: 'account.amazonaws.com',
        });
        const response = await organizationsClient.send(command);
        console.log('Trusted access enabled:', response);
      } catch (error) {
        console.error('Error enabling trusted access:', error);
      }

      // update STS tokens to V2
      await Promise.all(accountIds.map(accountId => setStsTokenPreferences(accountId, globalRegion)));

      return {
        IsComplete: false,
      };

    case 'Delete':
      // Do Nothing
      return {
        IsComplete: true,
      };
  }
}
