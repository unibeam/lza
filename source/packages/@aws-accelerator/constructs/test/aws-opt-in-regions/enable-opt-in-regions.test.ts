/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { OptInRegions } from '../../lib/aws-opt-in-regions/enable-opt-in-regions';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(EnableOptInRegions): ';

const app = new cdk.App();

// Create stack for native Cfn construct
const env = { account: '333333333333', region: 'us-east-1' };
const stack = new cdk.Stack(app, 'Stack', { env: env });

new OptInRegions(stack, 'OptInRegions', {
  kmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  logRetentionInDays: 3653,
  managementAccountId: '333333333333',
  accountIds: ['333333333333', '444444444444', '555555555555', '666666666666'],
  homeRegion: 'us-east-1',
  enabledRegions: ['ca-west-1', 'eu-south-2'],
  globalRegion: 'us-east-1',
});

/**
 * ConfigServiceTags construct test
 */
describe('EnableOptInRegions', () => {
  snapShotTest(testNamePrefix, stack);
});
