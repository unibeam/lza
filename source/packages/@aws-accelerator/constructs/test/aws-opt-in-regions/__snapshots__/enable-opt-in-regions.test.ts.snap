// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnableOptInRegions Construct(EnableOptInRegions):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::333333333333:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "OptInRegions8611CE24": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "OptInRegionsOptInRegionsIsCompleteLogGroupF6E1E28E",
        "OptInRegionsOptInRegionsOnEventLogGroupC61703D9",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsProviderframeworkonEventA805CEDE",
            "Arn",
          ],
        },
        "props": {
          "accountIds": [
            "333333333333",
            "444444444444",
            "555555555555",
            "666666666666",
          ],
          "enabledRegions": [
            "ca-west-1",
            "eu-south-2",
          ],
          "globalRegion": "us-east-1",
          "homeRegion": "us-east-1",
          "managementAccountId": "333333333333",
        },
      },
      "Type": "Custom::OptInRegions",
      "UpdateReplacePolicy": "Delete",
    },
    "OptInRegionsOptInRegionsIsComplete20A899A2": {
      "DependsOn": [
        "OptInRegionsOptInRegionsIsCompleteServiceRoleDefaultPolicy5BCB1377",
        "OptInRegionsOptInRegionsIsCompleteServiceRole88C6992C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Opt-in Regions isComplete handler",
        "Handler": "index.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "Role": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsIsCompleteServiceRole88C6992C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OptInRegionsOptInRegionsIsCompleteLogGroupF6E1E28E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "OptInRegionsOptInRegionsIsComplete20A899A2",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "OptInRegionsOptInRegionsIsCompleteServiceRole88C6992C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OptInRegionsOptInRegionsIsCompleteServiceRoleDefaultPolicy5BCB1377": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "account:ListRegions",
                "account:EnableRegion",
                "account:GetRegionOptStatus",
              ],
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "AccountOperations",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OptInRegionsOptInRegionsIsCompleteServiceRoleDefaultPolicy5BCB1377",
        "Roles": [
          {
            "Ref": "OptInRegionsOptInRegionsIsCompleteServiceRole88C6992C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OptInRegionsOptInRegionsOnEventDF9D2B80": {
      "DependsOn": [
        "OptInRegionsOptInRegionsOnEventServiceRoleDefaultPolicy3F5DA2D9",
        "OptInRegionsOptInRegionsOnEventServiceRoleD4B35886",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Opt-in Regions onEvent handler",
        "Handler": "index.handler",
        "KmsKeyArn": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "Role": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsOnEventServiceRoleD4B35886",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OptInRegionsOptInRegionsOnEventLogGroupC61703D9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "OptInRegionsOptInRegionsOnEventDF9D2B80",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "OptInRegionsOptInRegionsOnEventServiceRoleD4B35886": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OptInRegionsOptInRegionsOnEventServiceRoleDefaultPolicy3F5DA2D9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "organizations:EnableAWSServiceAccess",
                "iam:GetAccountSummary",
                "iam:SetSecurityTokenServicePreferences",
              ],
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "serviceAccessAndToken",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OptInRegionsOptInRegionsOnEventServiceRoleDefaultPolicy3F5DA2D9",
        "Roles": [
          {
            "Ref": "OptInRegionsOptInRegionsOnEventServiceRoleD4B35886",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OptInRegionsOptInRegionsProviderframeworkisCompleteA55ADDD8": {
      "DependsOn": [
        "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRoleDefaultPolicyFC3390B1",
        "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRole3768088A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - isComplete (Stack/OptInRegions/OptInRegionsProvider)",
        "Environment": {
          "Variables": {
            "USER_IS_COMPLETE_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "OptInRegionsOptInRegionsIsComplete20A899A2",
                "Arn",
              ],
            },
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "OptInRegionsOptInRegionsOnEventDF9D2B80",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.isComplete",
        "Role": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRole3768088A",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRole3768088A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRoleDefaultPolicyFC3390B1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsOnEventDF9D2B80",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsOnEventDF9D2B80",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsIsComplete20A899A2",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsIsComplete20A899A2",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRoleDefaultPolicyFC3390B1",
        "Roles": [
          {
            "Ref": "OptInRegionsOptInRegionsProviderframeworkisCompleteServiceRole3768088A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OptInRegionsOptInRegionsProviderframeworkonEventA805CEDE": {
      "DependsOn": [
        "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleDefaultPolicy6A8AF090",
        "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleE512B52D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Stack/OptInRegions/OptInRegionsProvider)",
        "Environment": {
          "Variables": {
            "USER_IS_COMPLETE_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "OptInRegionsOptInRegionsIsComplete20A899A2",
                "Arn",
              ],
            },
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "OptInRegionsOptInRegionsOnEventDF9D2B80",
                "Arn",
              ],
            },
            "WAITER_STATE_MACHINE_ARN": {
              "Ref": "OptInRegionsOptInRegionsProviderwaiterstatemachine380C9833",
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleE512B52D",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleDefaultPolicy6A8AF090": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsOnEventDF9D2B80",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsOnEventDF9D2B80",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsIsComplete20A899A2",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsIsComplete20A899A2",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "OptInRegionsOptInRegionsProviderwaiterstatemachine380C9833",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleDefaultPolicy6A8AF090",
        "Roles": [
          {
            "Ref": "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleE512B52D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OptInRegionsOptInRegionsProviderframeworkonEventServiceRoleE512B52D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OptInRegionsOptInRegionsProviderframeworkonTimeout2FB7EB0E": {
      "DependsOn": [
        "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleDefaultPolicyC43EC700",
        "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleC3734E22",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onTimeout (Stack/OptInRegions/OptInRegionsProvider)",
        "Environment": {
          "Variables": {
            "USER_IS_COMPLETE_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "OptInRegionsOptInRegionsIsComplete20A899A2",
                "Arn",
              ],
            },
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "OptInRegionsOptInRegionsOnEventDF9D2B80",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onTimeout",
        "Role": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleC3734E22",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleC3734E22": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleDefaultPolicyC43EC700": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsOnEventDF9D2B80",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsOnEventDF9D2B80",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsIsComplete20A899A2",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsIsComplete20A899A2",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleDefaultPolicyC43EC700",
        "Roles": [
          {
            "Ref": "OptInRegionsOptInRegionsProviderframeworkonTimeoutServiceRoleC3734E22",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OptInRegionsOptInRegionsProviderwaiterstatemachine380C9833": {
      "DependsOn": [
        "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleDefaultPolicy551C77D5",
        "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleB3284282",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"framework-isComplete-task","States":{"framework-isComplete-task":{"End":true,"Retry":[{"ErrorEquals":["States.ALL"],"IntervalSeconds":120,"MaxAttempts":15,"BackoffRate":1}],"Catch":[{"ErrorEquals":["States.ALL"],"Next":"framework-onTimeout-task"}],"Type":"Task","Resource":"",
              {
                "Fn::GetAtt": [
                  "OptInRegionsOptInRegionsProviderframeworkisCompleteA55ADDD8",
                  "Arn",
                ],
              },
              ""},"framework-onTimeout-task":{"End":true,"Type":"Task","Resource":"",
              {
                "Fn::GetAtt": [
                  "OptInRegionsOptInRegionsProviderframeworkonTimeout2FB7EB0E",
                  "Arn",
                ],
              },
              ""}}}",
            ],
          ],
        },
        "LoggingConfiguration": {
          "Destinations": [
            {
              "CloudWatchLogsLogGroup": {
                "LogGroupArn": {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsProviderwaiterstatemachineLogGroupDDB06910",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": false,
          "Level": "ERROR",
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleB3284282",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "OptInRegionsOptInRegionsProviderwaiterstatemachineLogGroupDDB06910": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/vendedlogs/states/waiter-state-machine-",
              {
                "Ref": "OptInRegionsOptInRegionsProviderframeworkisCompleteA55ADDD8",
              },
              "-c841d4c9c0fdb8d6f91953b31a208bd1b2bcd0c384",
            ],
          ],
        },
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleB3284282": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleDefaultPolicy551C77D5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsProviderframeworkisCompleteA55ADDD8",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsProviderframeworkisCompleteA55ADDD8",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OptInRegionsOptInRegionsProviderframeworkonTimeout2FB7EB0E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OptInRegionsOptInRegionsProviderframeworkonTimeout2FB7EB0E",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:CreateLogStream",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutLogEvents",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleDefaultPolicy551C77D5",
        "Roles": [
          {
            "Ref": "OptInRegionsOptInRegionsProviderwaiterstatemachineRoleB3284282",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
